lib/jortt/client/resource.rb

```diff
@@ -12,20 +33,27 @@ def initialize(token)
       private
 
       def get(path, params = {})
-        handle_response token.get(path, params: params)
+        handle_response { token.get(path, params: params) }
       end
 
       def post(path, params = {})
-        handle_response token.post(path, params: params)
+        handle_response { token.post(path, params: params) }
       end
 
       def put(path, params = {})
-        handle_response token.put(path, params: params)
+        handle_response { token.put(path, params: params) }
+      end
+
+      def delete(path)
+        handle_response { token.delete(path) }
       end
 
-      def handle_response(response)
+      def handle_response(&block)
+        response = yield
         return true if response.status == 204
         response.parsed.fetch('data')
+      rescue OAuth2::Error => e
+        raise ResponseError.new(e.response)
```

Comment:
Yes, you could also decouple the response from the error. Just rename `ResponseError` to `Error` and use a factory method to 'create' an `Error` instance from a response like so:

```
class Error
  def initialize(code, key, etc...)
    @code = code
  end

  def self.from_response(response)
    Error.new(
      response.parsed.dig('error', 'code'),
      etc...
    )
  end
end
```

Ideally also add the HTTP status code to the `Error`.

Reply:
In this case I feel like the coupling is inherent as we want to pass the error details on to the developer for debugging purposes.